gpaw: init at 22.8.0

sheepforce opened the pull request:
###### Description of changes

Part of the ongoing effort to upstream packages from the [NixOS-Qchem overlay](https://github.com/markuskowa/NixOS-QChem) (https://github.com/markuskowa/NixOS-QChem/issues/146). This adds the [GPAW](https://wiki.fysik.dtu.dk/gpaw/) package for condensed phase DFT.

The GPAW dependency LibVdwXC has also been added and requires a commit from master shortly after the latest stable release to work with GPAW.

GPAW requires an MPI-capable FFTW to work in parallel which is overriden on the fly. Not sure if this is the preferred option to do these things.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

markuskowa commented on pkgs/development/libraries/libvdwxc/default.nix:
```diff
@@ -0,0 +1,54 @@
+stdenv.mkDerivation rec {
+  pname = "libvdwxc";
+  # Stable version has non-working MPI detection.
+  version = "24.02.2020";
```
```suggestion
  version = "unstable-2020-02-24";
```

markuskowa commented on pkgs/development/libraries/libvdwxc/default.nix:
```diff
@@ -0,0 +1,54 @@
+
+  configureScript = "../configure";
+
+  src = fetchFromGitLab  {
```
Nitpick: moveup, put after "version"?

markuskowa commented on pkgs/development/python-modules/gpaw/default.nix:
```diff
@@ -0,0 +1,126 @@
+  passthru = { inherit mpi; };
+
+  meta = with lib; {
+    description = "DFT and beyond within the projector-augmented wave method";
```
```suggestion
    description = "Density functional theory and beyond within the projector-augmented wave method";
```
Avoid confusion with "discrete Fourier transform"?

risicle commented on pkgs/development/libraries/libvdwxc/default.nix:
```diff
@@ -0,0 +1,54 @@
+  ];
+
+  meta = with lib; {
+    description = "Portable C library of density functionals with van der Waals interactions for density functional theory";
```
`pkgs/development/libraries` these days has a `science` subdir that this probably belongs in.

risicle commented on pkgs/development/libraries/libvdwxc/default.nix:
```diff
@@ -0,0 +1,54 @@
+    rev = "92f4910c6ac88e111db2fb3a518089d0510c53b0";
+    sha256= "1c7pjrvifncbdyngs2bv185imxbcbq64nka8gshhp8n2ns6fids6";
+  };
+
```
```suggestion
  doCheck = true;
```

There isn't a lot here test-wise, but it's something.

risicle commented on pkgs/top-level/python-packages.nix:
```diff
@@ -3882,6 +3882,10 @@ in {
   gpapi = callPackage ../development/python-modules/gpapi { };
 
+  gpaw = callPackage ../development/python-modules/gpaw {
+    fftw = pkgs.fftw.override { enableMpi = true; };
```
If it's important that `gpaw` gets the *same* `fftw` as `libvdwxc` it might be an idea to add it to `libvdwxc`'s `passthru` and get it from there.

sheepforce replied:
Good suggestion, has been done :)

markuskowa replied:
Instead of using `passthru`, wouldn't it be better to put it `propgatedBuildInputs`?

sheepforce replied:
Hm, I am not sure about this one. The point here is that libvdwxc requires FFTW, not necessarily with MPI support on its own. GPAW however does require FFTW with MPI support and the libvdwxc that GPAW uses then also needs fftw with mpi support.

But an hypothetical other package (not that I am aware of one) that utilises libvdwxc that does not require fftw directly has no use for its propagated fftw.

Maybe a better alternative is to make MPI optional in libvdwxc 
```nix
{..., 
fftw,
mpi,
enableMpi ? true
}:

let fftwMpi = fftw.override { enableMpi = true; };

in stdenv.mkDerivation {
  ...
  buildInputs = [ ... ] ++ if enableMpi then [fftwMpi mpi] else [fftw];
}
```

and in gpaw
```nix
{..., 
fftw,
libvdwxc
}:
let libvdwxcMpi = libvdwxc.override { enableMpi = true; };
    fftwMpi = fftw.override { enableMpi = true; };
...
```

Also not optimal but maybe better?

markuskowa replied:
I think that `propagatedBuildInputs` could be the simpler solution here(?):
* Yes, a package that uses `libvdwxc` does not need `fftw` will be offered `fftw`. However, this should not hurt either.
* With `fftwMpi` in `propagatedBuildInputs` you can even remove `fftw` altogether from the `gpaw` inputs, and it is guaranteed that `gpaw` will get the right `fftw` version.
* Using `pasthru` is more complex: anyone building a new package needs to lookup `libvdwxc`'s passthru flags to get the `fftw` right.

Btw.: I think we can even define a top-level `fftwMpi` attribute now.

sheepforce replied:
>  I think we can even define a top-level fftwMpi attribute now.

I have added one :)

I guess multiple options are somehow valid. For now I've implemented the `propagatedBuildInputs` version.

sheepforce committed:
fftwMpi: add attribute for MPI-enabled fftw

sheepforce committed:
gpaw: init at 22.8.0


gpaw: use passthru fftw from libvdwxc


gpaw: fixes

sheepforce committed:
libvdwxc: init at 24.02.2020


libvdwxc: passthru mpi and do checks


libvdwxc: formatting


libvdwxc: fix evaluation
